docs/dxp/latest/en/process-automation/forms/developer-guide/writing-a-custom-data-provider/resources/liferay-b4d8.zip/b4d8-impl/src/main/java/com/acme/b4d8/dynamic/data/mapping/data/provider/internal/B4D8DataProvider.java

```diff
@@ -0,0 +1,187 @@
+			return doGetData(ddmDataProviderRequest);
+		}
+		catch (Exception exception) {
+			throw new DDMDataProviderException(exception);
+		}
+	}
+
+	@Override
+	public Class<?> getSettings() {
+		return ddmDataProviderSettingsProvider.getSettings();
+	}
+
+	protected DDMDataProviderResponse createDDMDataProviderResponse(
+			B4D8DataProviderSettings b4d8DataProviderSettings,
+			Document document)
+		throws Exception {
+
+		DDMDataProviderResponse.Builder builder =
+			DDMDataProviderResponse.Builder.newBuilder();
+
+		for (DDMDataProviderOutputParametersSettings outputParameterSettings :
+				b4d8DataProviderSettings.outputParameters()) {
+
+			String id = outputParameterSettings.outputParameterId();
+			String type = outputParameterSettings.outputParameterType();
+
+			NodeList nodeList = document.getElementsByTagName(
+				outputParameterSettings.outputParameterPath());
+
+			if (Objects.equals(type, "list")) {
+				List<KeyValuePair> keyValuePairs = new ArrayList<>();
+
+				for (int i = 0; i < nodeList.getLength(); i++) {
+					Node node = nodeList.item(i);
+
+					String nodeTextContent = node.getTextContent();
+
+					keyValuePairs.add(
+						new KeyValuePair(nodeTextContent, nodeTextContent));
+				}
+
+				builder = builder.withOutput(id, keyValuePairs);
+			}
+			else if (Objects.equals(type, "number")) {
+				Node node = nodeList.item(0);
+
+				NumberFormat numberFormat = NumberFormat.getInstance();
+
+				builder = builder.withOutput(
+					id, numberFormat.parse(node.getTextContent()));
+			}
+			else if (Objects.equals(type, "text")) {
+				Node node = nodeList.item(0);
+
+				String nodeTextContent = node.getTextContent();
+
+				builder = builder.withOutput(id, nodeTextContent);
+			}
+		}
+
+		return builder.build();
+	}
+
+	protected DDMDataProviderResponse doGetData(
+			DDMDataProviderRequest ddmDataProviderRequest)
+		throws Exception {
+
+		Optional<DDMDataProviderInstance> ddmDataProviderInstance =
+			fetchDDMDataProviderInstance(
+				ddmDataProviderRequest.getDDMDataProviderId());
+
+		B4D8DataProviderSettings b4d8DataProviderSettings =
+			ddmDataProviderInstanceSettings.getSettings(
+				ddmDataProviderInstance.get(), B4D8DataProviderSettings.class);
+
+		HttpRequest httpRequest = HttpRequest.get(
+			b4d8DataProviderSettings.url());
+
+		httpRequest.trustAllCerts(true);
+
+		HttpResponse httpResponse = httpRequest.send();
+
+		httpResponse.charset("UTF-8");
+
+		Document document = _convertXMLStringToDocument(
+			httpResponse.bodyText());
+
+		return createDDMDataProviderResponse(
+			b4d8DataProviderSettings, document);
+	}
+
+	protected Optional<DDMDataProviderInstance> fetchDDMDataProviderInstance(
+			String ddmDataProviderInstanceId)
+		throws Exception {
+
+		DDMDataProviderInstance ddmDataProviderInstance =
+			ddmDataProviderInstanceService.fetchDataProviderInstanceByUuid(
+				ddmDataProviderInstanceId);
+
+		if ((ddmDataProviderInstance == null) &&
+			Validator.isNumber(ddmDataProviderInstanceId)) {
```

Comment:
What do you do if `ddmDataProviderInstanceId` isn't a number?

Reply:
In this case the `return Optional.ofNullable(ddmDataProviderInstance);` will return a `Optional.empty`, since `ddmDataProviderInstance` is null.